### PR TITLE
Use wxOverlay to show sash feedback in wxAUI and in wxSplitterWindow

### PR DESCRIPTION
**Pull request by AliKet**
See #24177

Tested under: `wxMSW`, `wxGTK3` (**Wayland** & **X11**) and `wxQt`

Not Tested under `wxMac` though!

### CODE REVIEW

**Comment by AliKet** [comment]
May conflict with #24166

**Comment by AliKet** [comment]
I just tested this under `wxGTK2` and didn't work unfortunately.

**Comment by AliKet** [comment]
@vadz @paulcor  If I apply this patch:
```
diff --git a/src/gtk/dcclient.cpp b/src/gtk/dcclient.cpp
index 69478afc92..c2ed3f7afa 100644
--- a/src/gtk/dcclient.cpp
+++ b/src/gtk/dcclient.cpp
@@ -452,6 +452,11 @@ void wxWindowDCImpl::SetUpDC( bool isMemDC )
     gdk_gc_set_clip_rectangle( m_brushGC, nullptr );
     gdk_gc_set_clip_rectangle( m_textGC, nullptr );
     gdk_gc_set_clip_rectangle( m_bgGC, nullptr );
+
+    gdk_gc_set_subwindow( m_penGC, GDK_INCLUDE_INFERIORS );
+    gdk_gc_set_subwindow( m_brushGC, GDK_INCLUDE_INFERIORS );
+    gdk_gc_set_subwindow( m_textGC, GDK_INCLUDE_INFERIORS );
+    gdk_gc_set_subwindow( m_bgGC, GDK_INCLUDE_INFERIORS );
 }
 
 void wxWindowDCImpl::DoGetSize( int* width, int* height ) const
```
then this PR works correctly under wxGTK2.

**Comment by vadz** [comment]
Sorry, I've just realized that I perhaps shouldn't have merged #24166 before this one. But I can resolve the conflicts, however I'd like to understand the situation with wxGTK2: can we use `wxOverlay` for it or not, finally? I'd like to minimize time spent on it, but AFAIK we don't plan to drop support for it yet, so we still need this to work in at least some way there...

**Comment by AliKet** [comment]
> Sorry, I've just realized that I perhaps shouldn't have merged #24166 before this one. But I can resolve the conflicts, however I'd like to understand the situation with wxGTK2: can we use `wxOverlay` for it or not, finally? I'd like to minimize time spent on it, but AFAIK we don't plan to drop support for it yet, so we still need this to work in at least some way there...

`wxAUI` and `wxSplitterWindow` now work correctly under all supported platforms (can not test under **wxMac** !!)

**Comment by AliKet** [comment]
I also managed to get transparent hint to work with `wxOverlay` (generic and native), so `wxPseudoTransparentFrame` (with all its complications and limitation) is no longer needed.
I just want to know if it's ok to push the changes (depend on the changes done here) here or should I wait until this is approved and merged?

as always I need someone else to test my changes under **wxOSX**. TIA!

File: src/msw/overlay.cpp
Comment (vadz) [inline src/msw/overlay.cpp:111]: Why can't we always just use `GetScreenRect()` here? Is it just because we really need client size for `wxClientDC` but full window size otherwise? If so, we could still use something like

```c++
m_rect = wxRect(m_window->GetScreenPosition(), dc->GetSize());
```

I think, without any dynamic casts.

File: src/common/overlaycmn.cpp
Comment (vadz) [inline src/common/overlaycmn.cpp:143]: I would just put this in `Init()` itself, i.e. avoid defining the extra function which is used for wxGTK2 only but just inlined its body there inside `#if GTK2` test.

The rationale is that it's not obvious when seeing the call to `wxDontClipSubWindows()` that it only does something for GTK2, and it would be helpful to see it immediately.

This would also be a bit shorter.

File: src/generic/splitter.cpp
Comment (vadz) [inline src/generic/splitter.cpp:86]: We should extract this function somewhere and probably cache the bitmap because even if it's just a 2x2 one, I am still not sure we want to recreate it on every repaint.

The trouble is that we need to clean it up in some module and I don't see anything obviously appropriate. Perhaps we could make this a stock brush? Then it would be cleaned up by `wxStockGDI::DeleteAll()`.

But at least let's factor out this function into some private header, it's not quite trivial and it might need to be changed later (for dark mode or high DPI or I don't know what else) and so I'd rather not have 2 copies of it.

**Review by vadz — COMMENTED** [review]
Thanks a lot, this is much better and I'd really like to merge it but I'd like to avoid duplicating the function for creating the stipple bitmap. Will you be able to change it or should I do it myself?

**Comment by vadz** [comment]
> I also managed to get transparent hint to work with `wxOverlay` (generic and native), so `wxPseudoTransparentFrame` (with all its complications and limitation) is no longer needed.

Excellent!

> I just want to know if it's ok to push the changes (depend on the changes done here) here or should I wait until this is approved and merged?

Whatever you prefer. I.e. I'm fine with adding them here.

> as always I need someone else to test my changes under **wxOSX**. TIA!

I'll do it before merging, so from this point of view it would be better to remove `wxPseudoTransparentFrame` now, as I could test this too then.

Thanks again!

File: src/msw/overlay.cpp
Comment (AliKet) [reply src/msw/overlay.cpp:111]: When `wxClientDC` is used we want the overlay window to be positioned over the client area of the window. `m_window->GetScreenPosition()` returns the upper left corner of the window and not the top left position of the client area.

The overlay window positioned at `m_window->GetScreenPosition()` (**incorrect**): 

overlay_incorrect

The overlay window positioned at `m_window->GetClientRect()` (**correct**):

overlay_correct

Comment (vadz) [reply src/msw/overlay.cpp:111]: Ah, I see, thanks for the explanation!

I'd still like to avoid dynamic cast, for all the usual reasons. As always, we can just add a virtual function instead and while it's a bit ugly to have window-oriented function in `wxDC`, we do already have `GetWindow()` there, so we could also add some `GetOriginInWindow()` or `GetWindowOffset()` or whatever else you'd prefer that would return `wxPoint(0, 0)` by default but would be overridden to return `GetClientAreaOrigin()` in `wxClientDCImpl`.

Do you see any problems with doing it like this? Thanks again!

File: src/generic/splitter.cpp
Comment (AliKet) [reply src/generic/splitter.cpp:86]: Updating the documentation and this refactoring are still not done yet. if you don't mind me asking you to do that when testing under wxOSX, or fixing any regressions introduced (which I haven't noticed any!). TIA!

**Review by vadz — COMMENTED** [review]
Thanks again for all your work here, this looks really nice but I just have a couple of questions about opacity semantics under MSW, notably its default behaviour, could you please check this? TIA!

File: src/msw/overlay.cpp
Comment (AliKet) [reply src/msw/overlay.cpp:141]: This change is mainly to prevent the overlay window from being placed on top of/covering any floating window.

**Before:**

overlay_bad

**After:**

overlay_good

Unfortunately, this breaks rubberbanding in the drawing sample. That is, the rubberband no longer appears there! What actually solves both problems is doing the following:
```
    const auto win = wxGetTopLevelParent(m_window);
    HWND hWndInsertAfter = ::GetNextWindow(GetHwndOf(win), GW_HWNDPREV);
    m_overlayWindow = wxCreateOverlayWindow(m_rect, m_alpha, hWndInsertAfter);
```
Which I'll commit if you don't see any problem in doing so (of course) ?

File: include/wx/dc.h
Comment (AliKet) [reply include/wx/dc.h:722]: Documented now!

File: src/aui/framemanager.cpp
Comment (AliKet) [reply src/aui/framemanager.cpp:571]: Updated.

**Comment by AliKet** [comment]
@vadz I see that [`wxSplitterWindow::AlwaysUsesLiveUpdate`](https://docs.wxwidgets.org/latest/classwx_splitter_window.html#a7a377a089bb00011d00f233b055a5f8f)  is only added in 3.3.0.
Should we remove it completely or update its documentation like we did with `wxAuiManager::AlwaysUsesLiveResize`?

**Comment by vadz** [comment]
> @vadz I see that [`wxSplitterWindow::AlwaysUsesLiveUpdate`](https://docs.wxwidgets.org/latest/classwx_splitter_window.html#a7a377a089bb00011d00f233b055a5f8f) is only added in 3.3.0. Should we remove it completely or update its documentation like we did with `wxAuiManager::AlwaysUsesLiveResize`?

Yes, I think it should be fine to remove it if we never do this any longer. Thanks!

File: interface/wx/overlay.h
Comment (paulcor) [inline interface/wx/overlay.h:37]: That's not my understanding of how 'wxOverlay` is supposed to work. It's expected that calling `wxDCOverlay::Clear()` will make the overlay transparent.

Comment (AliKet) [reply interface/wx/overlay.h:37]: In fact, it works as you expect. Just the wording is misleading and should say something like this instead:
> Drawing on the overlay window is opaque by default under wxMSW...

because this is how `SetLayeredWindowAttributes` works. i.e. it can only apply a constant alpha on the drawing with no alpha blending at all.

As pictures speak louder than words, this is what I mean with rubberbanding as example:

- Rubberbanding (outline) with `SetLayeredWindowAttributes` + `wxALPHA_OPAQUE` (default):

overlay_opaque_outline

- Rubberbanding (filled) with `SetLayeredWindowAttributes` + `wxALPHA_OPAQUE` :

overlay_opaque

- Rubberbanding (filled) with `SetLayeredWindowAttributes` + alpha (128): 

overlay_translucent

- Rubberbanding (filled) with `UpdateLayeredWindow` + wxGCDC: 

overlay_translucent_gc

File: include/wx/dc.h
Comment (AliKet) [reply include/wx/dc.h:183]: I'll remove this commit then.

BTW, `wxOverlay` and `wxDCOverlay` documentations need to be updated. Also mentione that `wxClientDC` is the only supported DC that should be used to draw on the overlay!

Hope I understood you correctly!!

File: src/generic/splitter.cpp
Comment (AliKet) [reply src/generic/splitter.cpp:81]: Would it be better to do the refactoring in a seperate PR instead?

**Review by vadz — APPROVED** [review]
Thanks a lot once again for all your work on this!

I've finally tested this under Mac and I found one problem there: the overlay contents continued to be displayed after `Reset()` (although it disappeared as soon as the window was redrawn). I couldn't find any better way to fix this than the brute force approach implemented in 4261e1d07d52c0703a5adf4a436937eb0f8896e2 but at least with this it works for me.

@csomor If you have a moment to look at this commit and let me know if there is some simpler way of removing the overlay contents from screen, please let me know. FWIW I've tried calling `setNeedsDisplay:YES` but this didn't seem to help.

**Comment by vadz** [comment]
I'm merging this as it has been opened for a long time. We can almost surely improve things under Mac, but this can be done later.

Thanks again for all your efforts!

**Pull request merged**
